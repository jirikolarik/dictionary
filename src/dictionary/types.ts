export interface IDictionary {
  name: string;
  entries: {
    [key: string]: string;
  };
  ids: string[];
}

export interface IDictionaryState {
  active: string;
  data: {
    [key: string]: IDictionary;
  };
  ids: string[];
}

export interface IFormInput {
  name?: string;
  entries?: Array<{
    from?: string;
    to?: string;
  }>;
}

export interface IForm {
  name: string;
  entries: Array<{
    from: string;
    to: string;
  }>;
}

export enum DictionaryATs {
  ACTIVE = "ACTIVE",
  CREATE = "CREATE",
  UPDATE = "UPDATE",
  DELETE = "DELETE"
}
