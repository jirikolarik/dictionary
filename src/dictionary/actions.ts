import { action } from "typesafe-actions";
import uuid from "uuid/v1";
import { DictionaryATs, IForm } from "./types";

export const create = (payload: IForm) =>
  action(DictionaryATs.CREATE, { id: uuid(), ...payload });

export const update = (payload: { id: string; values: IForm }) =>
  action(DictionaryATs.UPDATE, { id: payload.id, ...payload.values });

export const remove = (id: string) => action(DictionaryATs.DELETE, id);

export const activate = (id: string) => action(DictionaryATs.ACTIVE, id);
