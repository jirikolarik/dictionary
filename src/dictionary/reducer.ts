import { dissocPath, mergeAll } from "ramda";
import { Reducer } from "redux";
import { ActionType } from "../types";
import { DictionaryATs, IDictionaryState } from "./types";

const id = "b3aa9c70-e35c-11e9-a411-516b6719292b";
export const initialState: IDictionaryState = {
  active: id,
  data: {
    [id]: {
      name: "First dictionary",
      entries: {
        Anthracite: "Dark Grey",
        "Midnight Black": "Black",
        "Mystic Silver": "Silver"
      },
      ids: ["Anthracite", "Midnight Black", "Mystic Silver"]
    }
  },
  ids: [id]
};

function standardise(values: Array<{ from: string; to: string }>) {
  return mergeAll(values.map(v => ({ [v.from]: v.to })));
}

const reducer: Reducer<IDictionaryState, ActionType> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case DictionaryATs.CREATE: {
      return {
        ...state,
        data: {
          ...state.data,
          [action.payload.id]: {
            name: action.payload.name,
            entries: standardise(action.payload.entries),
            ids: action.payload.entries.map(e => e.from)
          }
        },
        ids: [...state.ids, action.payload.id]
      };
    }
    case DictionaryATs.UPDATE: {
      return {
        ...state,
        data: {
          ...state.data,
          [action.payload.id]: {
            name: action.payload.name,
            entries: standardise(action.payload.entries),
            ids: action.payload.entries.map(e => e.from)
          }
        }
      };
    }
    case DictionaryATs.DELETE: {
      return {
        ...dissocPath(["data", action.payload], state),
        ids: state.ids.filter(id => id !== action.payload)
      };
    }
    case DictionaryATs.ACTIVE: {
      return {
        ...state,
        active: action.payload
      };
    }
    default: {
      return state;
    }
  }
};

export default reducer;
