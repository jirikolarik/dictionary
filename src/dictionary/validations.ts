import { findIndex, findLastIndex } from "ramda";
import { IFormInput } from "./types";

export function fieldRequired<T>(
  obj: T,
  key: keyof T,
  message = "Field is required"
) {
  return obj[key] === undefined ? { [key]: message } : {};
}

export function fieldHasDuplicate<T>(
  obj: T,
  key: keyof T,
  inArray: Array<string | undefined>,
  message = "This name is already in use"
) {
  return findIndex(a => a === obj[key])(inArray) !==
    findLastIndex(a => a === obj[key])(inArray)
    ? { [key]: message }
    : {};
}

export function validate(v: IFormInput) {
  const allKeys = (v.entries || []).map(e => e.from);
  return {
    ...fieldRequired(v, "name"),
    entries: (v.entries || []).map(e => ({
      ...fieldRequired(e, "from"),
      ...fieldRequired(e, "to"),
      ...fieldHasDuplicate(e, "from", allKeys)
    }))
  };
}
