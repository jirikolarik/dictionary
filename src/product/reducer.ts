import { Reducer } from "redux";
import { ActionType } from "../types";
import { IProductState } from "./types";

export const initialState: IProductState = {
  data: {
    "apple-iphone-6s": {
      color: "Anthracite",
      name: "Apple iPhone 6s",
      price: 769
    },
    "samsung-galaxy-s8": {
      color: "Midnight Black",
      name: "Samsung Galaxy S8",
      price: 569
    },
    "huawei-p9": {
      color: "Mystic Silver",
      name: "Huawei P9",
      price: 272
    }
  },
  ids: ["apple-iphone-6s", "samsung-galaxy-s8", "huawei-p9"]
};

const reducer: Reducer<IProductState, ActionType> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    default: {
      return state;
    }
  }
};

export default reducer;
