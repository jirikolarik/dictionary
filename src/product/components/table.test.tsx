import { mount } from "enzyme";
import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { initialState as dictionaryInitialState } from "../../dictionary/reducer";
import { initialState as productInitialState } from "../reducer";
import Table from "./table";

const mockStore = configureMockStore();

describe("Table", () => {
  it("should rename all entries", () => {
    const store = mockStore({
      dictionary: dictionaryInitialState,
      product: productInitialState
    });

    const wrapper = mount(
      <Provider store={store}>
        <Table />
      </Provider>
    );

    const text = wrapper.text();
    expect(wrapper.html()).toMatchSnapshot();
    expect(text).toContain("Name");
    expect(text).toContain("Color");
    expect(text).toContain("Price");
    expect(text).toContain("Apple iPhone 6s");
    expect(text).toContain("Dark Grey");
    expect(text).toContain("769");
    expect(text).toContain("Samsung Galaxy S8");
    expect(text).toContain("Black");
    expect(text).toContain("569");
    expect(text).toContain("Huawei P9");
    expect(text).toContain("Silver");
    expect(text).toContain("272");
  });
});
