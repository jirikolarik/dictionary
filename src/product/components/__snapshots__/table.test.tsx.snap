// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should rename all entries 1`] = `"<div data-baseweb=\\"table\\" role=\\"grid\\" aria-colcount=\\"3\\" aria-rowcount=\\"3\\" class=\\"\\"><div role=\\"row\\" class=\\"\\"><div role=\\"columnheader\\" class=\\"\\">Name</div><div role=\\"columnheader\\" class=\\"\\">Color</div><div role=\\"columnheader\\" class=\\"\\">Price</div></div><div role=\\"rowgroup\\" class=\\"\\"><div role=\\"row\\" class=\\"\\"><div role=\\"gridcell\\" class=\\"\\">Apple iPhone 6s</div><div role=\\"gridcell\\" class=\\"\\">Dark Grey</div><div role=\\"gridcell\\" class=\\"\\">CHF&nbsp;769.00</div></div><div role=\\"row\\" class=\\"\\"><div role=\\"gridcell\\" class=\\"\\">Samsung Galaxy S8</div><div role=\\"gridcell\\" class=\\"\\">Black</div><div role=\\"gridcell\\" class=\\"\\">CHF&nbsp;569.00</div></div><div role=\\"row\\" class=\\"\\"><div role=\\"gridcell\\" class=\\"\\">Huawei P9</div><div role=\\"gridcell\\" class=\\"\\">Silver</div><div role=\\"gridcell\\" class=\\"\\">CHF&nbsp;272.00</div></div></div></div>"`;
